src/atomic.cr

```diff
@@ -38,6 +69,79 @@ struct Atomic(T)
     Ops.cmpxchg(pointerof(@value), cmp.as(T), new.as(T), :sequentially_consistent, :sequentially_consistent)
   end
 
+  def compare_and_set(cmp : T, new : T, success_ordering : Ordering, failure_ordering : Ordering) : {T, Bool}
+    {% if compare_versions(Crystal::LLVM_VERSION, "13.0.0") >= 0 %}
+      # LLVM since 13.0.0 accepts any combination of success & failure ordering
+      case success_ordering
+        {% for s_ordering in Ordering.constants %}
+          in Ordering::{{s_ordering}}
+            case failure_ordering
+              {% for f_ordering in Ordering.constants %}
+                in Ordering::{{f_ordering}}
+                  Ops.cmpxchg(pointerof(@value), cmp.as(T), new.as(T),
+                              {{ s_ordering.stringify == "Relaxed" ? :monotonic : s_ordering.underscore.symbolize }},
+                              {{ f_ordering.stringify == "Relaxed" ? :monotonic : f_ordering.underscore.symbolize }})
+              {% end %}
+            end
+        {% end %}
+      end
+    {% else %}
+      # LLVM until 12.0.0 only accepts some combinations of success & failure ordering
```

Comment:
Sadly `{% f_ordering > s_ordering %}` compares the constant names (macro ids?) not the enum values, so that doesn't work (e.g `Acquire` < `Relaxed` is true :shrug:).